02-generics

SergeyMokin opened the pull request.

Commit:
done

AlexMokin commented on 02-Generics/Generics/Generics.cs:
```diff
@@ -91,10 +101,45 @@ public static class ArrayExtentions {
+            where T2: IComparable
+            where T3: IComparable
+        {
+                switch (sortedColumn)
```
do it without switch case statements

AlexMokin commented on 02-Generics/Generics/Generics.cs:
```diff
@@ -105,13 +150,22 @@ public static class ArrayExtentions {
+	public static class Singleton<T> where T: class, new() {
+        private static T instance = null;
+
+        public static T Instance
```
not thread safe

AlexMokin commented on 02-Generics/Generics/Generics.cs:
```diff
@@ -135,9 +189,19 @@ public static class FunctionExtentions {
+                    System.Diagnostics.Debug.WriteLine(exc);
+                }
+            }
+            throw new System.Net.WebException();
```
original exception has to be thrown

Commit:
I rewrite SortTupleArray without switch, but i have no idea how to do without if/else. I read some information about tuples, maybee i need to read something docs about generics. I'll do it tomorrow.
At Singleton class i realize static constructor, this will be run one time.
And at last task TimeoutSafeInvoke i all forgot about failed test, because i did this some time ago. Now I completed this task.

AlexMokin commented on 02-Generics/Generics/Generics.cs:
```diff
@@ -91,10 +101,27 @@ public static class ArrayExtentions {
+            {
+                throw new IndexOutOfRangeException();
+            }
+            Array.Sort(array, (x, y) => sortedColumn == 0
```
it's not better than with switch case, and it's more difficult to read

AlexMokin commented on 02-Generics/Generics/Generics.cs:
```diff
@@ -105,13 +132,24 @@ public static class ArrayExtentions {
-	}
+        static Singleton()
+        {
+            instance = new T();
```
make it lazy

Commit:
I rewrite Singleton class, make it lazy and delete static constructor, because by default Lazy  are thread-safe.
At SortTupleArray I use idea of Sergey Dubovich. I rly had no idea how to do this. I realized it differently, but the idea looked at him.

AlexMokin commented on 02-Generics/Generics/Generics.cs:
```diff
@@ -91,10 +101,38 @@ public static class ArrayExtentions {
+            where T3: IComparable
+        {
+            Func<Tuple<T1, T2, T3>, IComparable> column = tuple => {
+                if (sortedColumn == 0)
```
no ifs,  no switch case

Commit:
Done without switch/case

AlexMokin commented:
good
